Src/Support/GoogleApis.Auth.PlatformServices_Shared/OAuth2/LocalServerCodeReceiver.cs

```diff
@@ -0,0 +1,393 @@
+                return requestLine.ToString();
+            }
+
+            private Dictionary<string, string> ValidateAndGetRequestParams(string requestLine)
+            {
+                var requestLineParts = requestLine.Split(' ');
+                if (requestLineParts.Length != 3)
+                {
+                    throw new ServerException("Request line ill-formatted. Should be '<request-method> <request-path> HTTP/1.1'");
+                }
+                string requestVerb = requestLineParts[0];
+                if (requestVerb != "GET")
+                {
+                    throw new ServerException($"Expected  GET request, got '{requestVerb}'");
+                }
+                string requestPath = requestLineParts[1];
+                if (!requestPath.StartsWith(LoopbackCallbackPath))
+                {
+                    throw new ServerException($"Expected request path to start '{LoopbackCallbackPath}', got '{requestPath}'");
+                }
+                var pathParts = requestPath.Split('?');
+                if (pathParts.Length == 1)
+                {
+                    return new Dictionary<string, string>();
+                }
+                if (pathParts.Length != 2)
+                {
+                    throw new ServerException($"Expected a single '?' in request path, got '{requestPath}'");
+                }
+                var queryParams = pathParts[1];
+                var result = queryParams.Split(new[] { '&' }, StringSplitOptions.RemoveEmptyEntries).Select(param =>
+                {
+                    var keyValue = param.Split('=');
+                    if (keyValue.Length > 2)
+                    {
+                        throw new ServerException($"Invalid query parameter: '{keyValue}'");
+                    }
+                    var key = WebUtility.UrlDecode(keyValue[0]);
+                    var value = keyValue.Length == 2 ? WebUtility.UrlDecode(keyValue[1]) : "";
+                    return new { key, value };
+                }).ToDictionary(x => x.key, x => x.value);
+                return result;
+            }
+
+            private async Task WaitForAllHeaders(Func<Task<byte?>> getByte)
+            {
+                // Looking for an empty line, terminated by \r\n
+                int byteCount = 0;
+                int lineLength = 0;
+                char c0 = (char)0;
+                char c1 = (char)0;
+                while (true)
+                {
+                    if (byteCount > MaxHeadersLength)
+                    {
+                        throw new ServerException($"Headers too long: > {MaxHeadersLength} bytes.");
+                    }
+                    byte? b = await getByte();
+                    if (b == null)
+                    {
+                        throw new ServerException("Unexpected end of network stream waiting for headers.");
+                    }
+                    c0 = c1;
+                    c1 = (char)b;
```

Comment:
(Although we're not actually using the result here, right, other than to detect the line ending, right?)